controllers/remoteistio/remoteistio_controller.go

```diff
@@ -0,0 +1,315 @@
+}
+
+func validate(istio *v1alpha1.RemoteIstio) error {
+	if istio.Spec.Version == "" {
+		return reconciler.NewValidationError("spec.version not set")
+	}
+	if istio.Spec.Namespace == "" {
+		return reconciler.NewValidationError("spec.namespace not set")
+	}
+	return nil
+}
+
+func (r *Reconciler) reconcileActiveRevision(ctx context.Context, istio *v1alpha1.RemoteIstio) error {
+	values, err := revision.ComputeValues(
+		istio.Spec.Values, istio.Spec.Namespace, istio.Spec.Version,
+		r.DefaultProfile, istio.Spec.Profile,
+		r.ResourceDirectory, getActiveRevisionName(istio))
+	if err != nil {
+		return err
+	}
+
+	return revision.CreateOrUpdate(ctx, r.Client,
+		getActiveRevisionName(istio),
+		v1alpha1.IstioRevisionTypeRemote,
+		istio.Spec.Version, istio.Spec.Namespace, values,
+		metav1.OwnerReference{
+			APIVersion:         v1alpha1.GroupVersion.String(),
+			Kind:               v1alpha1.RemoteIstioKind,
+			Name:               istio.Name,
+			UID:                istio.UID,
+			Controller:         ptr.Of(true),
+			BlockOwnerDeletion: ptr.Of(true),
+		})
+}
+
+func getPruningGracePeriod(istio *v1alpha1.RemoteIstio) time.Duration {
+	strategy := istio.Spec.UpdateStrategy
+	period := int64(v1alpha1.DefaultRevisionDeletionGracePeriodSeconds)
+	if strategy != nil && strategy.InactiveRevisionDeletionGracePeriodSeconds != nil {
+		period = *strategy.InactiveRevisionDeletionGracePeriodSeconds
+	}
+	if period < v1alpha1.MinRevisionDeletionGracePeriodSeconds {
+		period = v1alpha1.MinRevisionDeletionGracePeriodSeconds
+	}
+	return time.Duration(period) * time.Second
+}
+
+func (r *Reconciler) getActiveRevision(ctx context.Context, istio *v1alpha1.RemoteIstio) (v1alpha1.IstioRevision, error) {
+	rev := v1alpha1.IstioRevision{}
+	err := r.Client.Get(ctx, getActiveRevisionKey(istio), &rev)
+	if err != nil {
+		return rev, fmt.Errorf("get failed: %w", err)
+	}
+	return rev, nil
+}
+
+func getActiveRevisionKey(istio *v1alpha1.RemoteIstio) types.NamespacedName {
+	return types.NamespacedName{
+		Name: getActiveRevisionName(istio),
+	}
+}
+
+func getActiveRevisionName(istio *v1alpha1.RemoteIstio) string {
+	var strategy v1alpha1.UpdateStrategyType
+	if istio.Spec.UpdateStrategy != nil {
+		strategy = istio.Spec.UpdateStrategy.Type
+	}
+
+	switch strategy {
+	default:
+		fallthrough
+	case v1alpha1.UpdateStrategyTypeInPlace:
+		return istio.Name
+	case v1alpha1.UpdateStrategyTypeRevisionBased:
+		return istio.Name + "-" + strings.ReplaceAll(istio.Spec.Version, ".", "-")
+	}
+}
+
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(controller.Options{
+			LogConstructor: func(req *reconcile.Request) logr.Logger {
+				log := mgr.GetLogger().WithName("ctrlr").WithName("istio")
+				if req != nil {
+					log = log.WithValues("Istio", req.Name)
+				}
+				return log
+			},
+		}).
+		For(&v1alpha1.RemoteIstio{}).
+		Owns(&v1alpha1.IstioRevision{}).
```

Comment:
Do you need a predicate filtering out revisions that aren't `type == Remote`? Probably it's an error to have both an `Istio` and `RemoteIstio` in your cluster at the same time.

Reply:
No. `Owns` is simply equivalent to `Watches(object, handler.EnqueueRequestForOwner([...], ownerType, OnlyControllerOwner()))`, so there's nothing wrong if two controllers own the same resource type. Both controllers will watch `IstioRevision` objects, but only one will actually process the watch event, depending on whether the `ownerReference` in the `IstioRevision` points to `Istio` or `RemoteIstio`, 

Also, I think it's a perfectly valid use-case to have both an `Istio` and `RemoteIstio` in the same cluster - it isn't any different from having two `Istio`s.